File: src/main/java/net/dean/jraw/paginators/SubredditSearchPaginator.java
package net.dean.jraw.paginators;

import net.dean.jraw.EndpointImplementation;
import net.dean.jraw.Endpoints;
import net.dean.jraw.RedditClient;
import net.dean.jraw.models.Listing;
import net.dean.jraw.models.Subreddit;

import java.util.HashMap;
import java.util.Map;

/**
 * This class provides a way to search for subreddits by both name and description
 */
public class SubredditSearchPaginator extends Paginator<Subreddit> {
    private String query;

    /**
     * Instantiates a new Paginator
     *
     * @param creator The RedditClient that will be used to send HTTP requests
     * @param query   What to search for
     */
    public SubredditSearchPaginator(RedditClient creator, String query) {
        super(creator, Subreddit.class);
        this.query = query;
    }

    @Override
    protected String getBaseUri() {
        return "/subreddits/search";
    }

    @Override
    @EndpointImplementation(Endpoints.SUBREDDITS_SEARCH)
    public Listing<Subreddit> next(boolean forceNetwork) throws IllegalStateException {
        return super.next(forceNetwork);
    }

    @Override
    protected Map<String, String> getExtraQueryArgs() {
        Map<String, String> args = new HashMap<>(super.getExtraQueryArgs());
        args.put("q", query);
        return args;
    }
}

File: lib/src/main/kotlin/net/dean/jraw/pagination/SubredditSearchPaginator.kt
package net.dean.jraw.pagination

import net.dean.jraw.RedditClient
import net.dean.jraw.http.HttpRequest
import net.dean.jraw.models.Subreddit
import net.dean.jraw.models.SubredditSearchSort
import net.dean.jraw.models.TimePeriod

/**
 * This specific paginator is dedicated to searching for subreddits. To create an instance, use
 * [RedditClient.searchSubreddits]
 */
class SubredditSearchPaginator private constructor(b: Builder) : DefaultPaginator<Subreddit>(
    reddit = b.reddit,
    sortingAlsoInPath = false,
    baseUrl = b.baseUrl,
    limit = b.limit,
    clazz = Subreddit::class.java,
    timePeriod = TimePeriod.ALL, // unused
    sorting = b.sorting
) {
    val query = b.query

    override fun createNextRequest(): HttpRequest.Builder {
        // Pretty much everything can be reused from DefaultPaginator
        val base = super.createNextRequest().build()

        // Add the 'q' query parameter
        return base.newBuilder()
            .configureUrl { it.setQueryParameter("q", query) }
    }

    /** Builder pattern implementation */
    class Builder(reddit: RedditClient, baseUrl: String = "/subreddits/search") : Paginator.Builder<Subreddit>(reddit, baseUrl, Subreddit::class.java) {
        internal var limit: Int = Paginator.DEFAULT_LIMIT
        internal var query: String = ""
        internal var sorting: SubredditSearchSort = SubredditSearchSort.RELEVANCE

        /**
         * Sets how many items to return in each page. The default is 25, and the maximum reddit allows is 100. If not
         * set, the built paginator will use the default value.
         *
         * @see Paginator.DEFAULT_LIMIT
         * @see Paginator.RECOMMENDED_MAX_LIMIT
         */
        fun limit(limit: Int): Builder { this.limit = limit; return this }

        /**
         * A term or terms to search for. Subreddits with this term in its name or description will be more likely to
         * appear in the results.
         */
        fun query(query: String): Builder { this.query = query; return this }

        /** How to sort the results. Defaults to [SubredditSearchSort.RELEVANCE]. */
        fun sorting(sorting: SubredditSearchSort): Builder { this.sorting = sorting; return this }

        override fun build(): SubredditSearchPaginator = SubredditSearchPaginator(this)
    }
}
